### ISSUE TITLE
[Federated Plugins] Provide a way to override a default/endorsed platform implementation

### ISSUE THREAD

**Issue by ryanheise** [flutter/flutter#80374]
## Use case

It would be desirable for a federated plugin to have a default implementation for iOS, either inside the plugin or an endorsed separate package, and to then allow apps to choose an alternative platform implementation to override that default. E.g. for just_audio, we will have a default AVQueuePlayer-based implementation, but an alternative AudioKit implementation and an alternative VLC implementation.

However, I don't think this override ability has been added to Flutter yet, although it was implied in the [design doc](https://docs.google.com/document/d/1LD7QjmzJZLCopUrFAAE98wOUQpjmguyGTN2wd_89Srs/edit#).

What happens instead is that the generated plugin registrant will automatically add both the default/endorsed plugin AND any additional plugin the app adds to their pubspec with the intention of overriding the default. Since both are listed, alphabetical order will dictate which one is effective.

## Proposal

Currently a platform implementation is endorsed by adding its dependency to the plugin's front-end and so it's going to get added to the generated plugin registrant regardless of any attempt to override it.

I'm not sure on the best solution, but maybe either the mechanism to endorse an implementation needs to be something other than a normal dependency, or otherwise pub should add some special cases to the code that generates the generated plugin registrant to detect overrides and avoids adding both the override and the default.

**Comment by stuartmorgan** [comment]
The PR for https://github.com/flutter/flutter/issues/52267 has logic to handle this for Dart plugin registration; we should be able to extend that same logic to any native plugin implementation.

**Comment by ryanheise** [comment]
By my understanding, that PR describes configuration changes to the plugins themselves, but I'm trying to understand what an app developer would need to configure to choose the overriding platform. Currently we would do something like this in the app's pubspec:

```yaml
dependencies:
  just_audio:
  just_audio_audiokit_ios:
```

where just_audio already includes a default iOS implementation, and each of those two packages would define . What mechanism controllable by the app itself can prevent the default implementation from loading? I assumed the plugin class is a property of the plugin rather than the app, but would the app need to define the plugin class for each override plugin it uses, in addition to adding the package dependency?

**Comment by stuartmorgan** [comment]
https://github.com/flutter/flutter/pull/79669/files#diff-68361f6fc9aafa5fa7de042eec4603dd4f321eb783a00edd57780ef78c1094bbR1051-R1060

**Comment by ryanheise** [comment]
```
/// Resolves the platform implementation for Dart-only plugins.
///
///   * If there are multiple direct pub dependencies on packages that implement the
///     frontend plugin for the current platform, fail.
///   * If there is a single direct dependency on a package that implements the
///     frontend plugin for the target platform, this package is the selected implementation.
///   * If there is no direct dependency on a package that implements the frontend
///     plugin for the target platform, and the frontend plugin has a default implementation
///     for the target platform the default implementation is selected.
///   * Else fail.
///
///  For more details, https://flutter.dev/go/federated-plugins.
```

Thanks, I think I understand the algorithm.

I was slightly confused by the wording above, because `just_audio` could be considered as providing a default iOS implementation itself, but it could also be considered as being a "direct pub dependency that implements the frontend plugin for the current platform".

**Comment by ryanheise** [comment]
Is the way to override an endorsed platform implementation now documented anywhere in the latest up-to-date documentation on federated plugins?

**Comment by stuartmorgan** [comment]
No; we're unlikely to formally document this until it works for native code as well, otherwise it will be very confusing.

**Comment by Gustl22** [comment]
I'm not too familiar with the whole federated & endorsed plugin management. Are there tendencies, how this will be approached?
You already wrote two solutions here, although they do not exclude each other and the second one does not solve the problem of this thread directly:
https://github.com/flutter/flutter/issues/81650#issuecomment-1634084576

1. I think the simplest way would be to have `dependency_overrides` which also allow packages with different name than the existing one for endorsed plugins, like
```yaml
dependency_overrides:
  my_plugin:
    path: ..
  my_plugin_android:
    path: ..\\..\\my_plugin_android_alternative
```
But this is kind of hacky, and also needs involving pub. This also doesn't work if dart code is involved as the package name differs, of course.


2. Maybe something like this is more practicable:
```yaml
flutter:
  plugin_overrides:
    my_plugin:
      platforms:
        android:
          default_package: my_plugin_android_alternative

dependencies:
  my_plugin_android_alternative: ^0.0.1
```

Which could also replace the proposal to exclude a plugin with [`plugin_exclusions`](https://github.com/flutter/flutter/issues/81650#issuecomment-1634084576) and would maybe more align the existing style:
```
flutter:
  plugin_overrides:
    my_plugin:
      platforms:
        android: {}
```

But still, I haven't looked into the code and don't know if it's actually possible to prevent a native plugin from registering, or it's just that the code isn't included in the build process at all. And then could just load the other plugin alternative or even no plugin.

**Comment by stuartmorgan** [comment]
> Are there tendencies, how this will be approached?

The comments above yours describe how this will be implemented.

> You already wrote two solutions here, although they do not exclude each other and the second one does not solve the problem of this thread directly: [#81650 (comment)](https://github.com/flutter/flutter/issues/81650#issuecomment-1634084576)

My comment there is not about solutions to this issue; it's about solving a different problem.

> But still, I haven't looked into the code and don't know if it's actually possible to prevent a native plugin from registering

[This](https://github.com/flutter/flutter/issues/80374#issuecomment-818973471) will, when implemented, cause the overridden implementation not to be registered.

**Comment by Gustl22** [comment]
> The comments above yours describe how this will be implemented.

Yes, now I got it. I just never saw how this actually is implemented in the `pubspec.yaml` . But I found also the [exact code line](https://github.com/flutter/flutter/blob/4b7b4d4176764ccbd2b4e407da56f3bb39805034/packages/flutter_tools/lib/src/flutter_plugins.dart#L1217) now.

So in dart it just works by adding a direct dependency to the developers application, it automatically detects if it has an implementation of an existing plugin package. And then prefer the direct dependency over the default one.
I was always the opinion one needs to actually declare the plugin override in the pubspec via the `flutter` params, but that isn't the case.

And it's already described in words [here](https://docs.flutter.dev/packages-and-plugins/developing-packages#non-endorsed-federated-plugin) for non-endorsed plugins:

```yaml
name: example
dependencies:
  my_plugin: ^0.0.1
  my_plugin_android_alternative: ^0.0.1
```

So the registration for the alternative plugin is already working, just the deregistration of the default one is not yet. I may have a deeper look into this.

**Comment by Gustl22** [comment]
I created a PR now (#137040) which aims to solve this issue. It's not ready yet (missing parts are listed in the PR description), but maybe one can comment, if it's leading in the right direction and one can have a look in the TODOs in the code.

Hope it can help to solve this issue :D
